Basic Heartbeat Job validation

andrewvc opened the pull request:
Takes over for #7551 which was getting quite long.

**Warning, I haven't gone over this line-by-line completely yet. It could probably use some more cleanup. Feedback on the skima DSL would be the most welcome sort of feedback at this point, as well as the general testing approach**

This PR seeks to establish a pattern for testing heartbeat jobs. It currently tests the HTTP and TCP jobs.

To do this, it made sense to validate event maps with a sort of schema library. I couldn't find one that did exactly what I wanted here, so I wrote one called `skima`.

One of the nicest things about the heartbeat architecture is the dialer chain behavior. It should be the case that any validated protocol using TCP (e.g. HTTP, TCP, Redis, etc.) has the *exact* same tcp metadata. 

To help make testing these properties easy `skima` lets users compose portions of a schema into a bigger one. In other words, you can say "An HTTP response should be a TCP response, with the standard monitor data added in, and also the special HTTP fields". Even having only written a handful of tests this has uncovered some inconsistencies there, where TCP jobs have a hostname, but HTTP ones do not. It's also uncovered that HTTP 5xx responses don't include a status code, while 2xx ones do.

Files worth looking at first:

1. HTTP Test: ([monitors/active/http/http_test.go](https://github.com/andrewvc/beats/blob/job_validation/heartbeat/monitors/active/http/http_test.go#L83))
2. TCP Test: ([monitors/active/tcp/tcp_test.go](https://github.com/andrewvc/beats/blob/job_validation/heartbeat/monitors/active/tcp/tcp_test.go#L31)
3. Skima Unit Tests: ([skima/core_test.go](https://github.com/andrewvc/beats/blob/job_validation/heartbeat/skima/core_test.go))

**Adendum, why I didn't choose other schema libs**

Well, this started off as being a very simple set of functions, and here we are. Still, that's not a good reason to write something new. So, let's compare:

1. [go-playground/go-validator](https://godoc.org/gopkg.in/go-playground/validator.v9) : awkward tag based syntax. A goal here was clarity, this isn't so easy to use.
2. [ozzo-validation](https://github.com/go-ozzo/ozzo-validation). Seems mostly targeted at web apps / forms. Doesn't really have a validation def that 'looks' like the end representation, which is a key goal of 'skima'
3. [go-validator/validator](https://github.com/go-validator/validator) once again, awkward tag based syntax.

ruflin commented on heartbeat/skima/schema_validation_results.go:
```diff
@@ -0,0 +1,56 @@
+
+// Valid returns true if there are no errors.
+func (r SchemaValidationResults) Valid() bool {
+	// TODO: this is a pretty slow way to do this.
```
As this is only for testing should not matter?

andrewvc replied:
Yep! I agree. N is pretty small.

One thing @urso did mention is we could eventually use this skima lib as the backend to a user-definable validator in YAML. I think it's important to call out slow functions so that if people use them in the future they're aware they may need optimization.

ruflin commented on heartbeat/testutil/hbtestutil.go:
```diff
@@ -0,0 +1,79 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package testutil
```
In other places we call this package `testing`

andrewvc replied:
++ will change

urso replied:
both `testing` and `testutil` are somewhat generic. We even have a `testing` package that is not used for testing but for config validation... I like what the go stdlib does. E.g. `httptest` or `iotest`. Depending how generic this package is `hbtest` might be fine :)

ruflin commented on vendor/vendor.json:
```diff
@@ -1361,6 +1361,12 @@
 			"versionExact": "v1.2.0"
 		},
+		{
+			"checksumSHA1": "wnEANt4k5X/KGwoFyfSSnpxULm4=",
```
I see this package gets added but I couldn't find where it's used?

andrewvc replied:
Ah, I did delete the test that used it, but I'm going to use it again soon. The existing tests should use this to `t.FailNow()` when a test server returns `err` or the like. Apologies. So, a common idiom would be:

```go
server, err := httptest.Server()
require.Nil(err)
```

urso replied:
We use assert where requires should be used much too often. +1 for adding it. But maybe in an extra PR, so we can use `requires` in the future.

andrewvc replied:
Hmmm, I just pushed new commits that make heavier use of `requires`. Do we still want to break that up into a separate PR? (I'm fine with that, not a big deal)

ruflin replied:
I'm good with keeping it in this PR as it's now actually used.

urso commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+		// It's a little weird, but is fairly efficient. We could stop using the flattened map as a datastructure, but
+		// that would add complexity elsewhere. Probably a good refactor at some point, but not worth it now.
+		validatedPaths := sort.StringSlice{}
+		for k, _ := range validatedResults {
```
btw. `gofmt -s` (-s == simplify) will correct this.

andrewvc commented on heartbeat/skima/core_test.go:
```diff
@@ -0,0 +1,216 @@
+package skima
+
+import (
+	"testing"
```
Ahhhh, I need to fix intellij to do the right thing. Sorry!

urso commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+		// and turn it into a sorted string array, then do a binary prefix search to determine if a subkey was tested.
+		// It's a little weird, but is fairly efficient. We could stop using the flattened map as a datastructure, but
+		// that would add complexity elsewhere. Probably a good refactor at some point, but not worth it now.
+		validatedPaths := sort.StringSlice{}
```
while it works a more idiomatic use of sort package is to use it only for sorting, not for it's types:

e.g.
```
validatedPaths := make([]string, len(validatedResults))
for k := range validatedResults {
```

urso commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+type Map map[string]interface{}
+
+// SchemaValidator is the result of Schema and is run against the map you'd like to test.
+type SchemaValidator func(common.MapStr) SchemaValidationResults
```
try to reduce stutter. The package kinda name resembles schema. For users of a package, a symbol defined in a package always has the name combined by package name + symbol name. E.g. the actual type name here is `skima.SchemaValidator`. Without stuttering we get `skima.Validator`.

Same for the validation results. The `SchemaValidationResults` will  be `skima.SchemaValidationResults`. Unless you have a more generic result type consider renaming this to `skima.Results`

andrewvc replied:
The confusing thing here is that there is a notion of Value validation vs. Schema validation in this package.

For values we have `IsDef`, `Checker`, and `ValueResult`
For map/schema validation we have `SchemaValidator` and `SchemaValidationResults`.

I'm thinking these names are a bit inconsistent, and I agree too long. WDYT about renaming them as indicated below?

`IsDef`->`IsDef`
`Checker`->`ValValidator`
`ValueResult`->`ValResult`
`SchemaValidator`->`Validator`
`SchemaValidationResults`->`Results`

urso replied:
SGTM. `ValueResult` is fine as well.

urso commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+// Strict is used when you want any unspecified keys that are encountered to be considered errors.
+func Strict(laxValidator SchemaValidator) SchemaValidator {
+	return func(actual common.MapStr) SchemaValidationResults {
+		validatedResults := laxValidator(actual)
```
everything is `validatedX` in here. Try to keep short names. E.g. code is understandable as well, by using `results, paths`.

andrewvc commented:
@ruflin thanks for the review.

Would love to get your input on this @jsoriano . I think the two packages have different aims. Am I correct in saying that `schema` seems to be more oriented around transformations? This package is solely focused on validation. There may be an opportunity here to unify stuff, we should sync on that.

urso commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+			path []string,
+			dottedPath string) {
+
+			_, validatedExactly := validatedResults[dottedPath]
```
as results is it's own type, these two lines are calling for a method so you can do: `if results.Has(dottedPath)`...

urso commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+import (
+	"strings"
+
+	"testing"
```
Do never import "testing" in packages that are not used for testing only. If skima is supposed to be used for testing only, then this should be reflected in the package path. Importaing "testing" has a many side-effects. E.g. global CLI flags we do not want (and that will clash with beats CLI flags) will be registered globally.

andrewvc replied:
Good to know! OK, I'll move the `Test` function out of `skima` into a new `skimat` support package. (does that naming work well? I'm trying to adjust to short var names in goland)

urso replied:
SGTM. `skimatest` will also do. One can still alias or wrap the function locally if required.

urso commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+		assert.True(t, vr.valid, msg)
+	})
+	return r
+}
```
How about changing this to `DetailedErrors` which returns an error wrapping multiple errors.

```
type multiErr []error

type (m *multiErr) Error() string {

andrewvc replied:
I like the idea, I'd still like to keep around the `map[string][]ValueResult` structure for future use, but for testing it is overkill.

andrewvc commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
```
+1 to do this for now. We might move it out later for runtime use, but that doesn't need to be the case for now.

urso commented on heartbeat/skima/walk.go:
```diff
@@ -0,0 +1,74 @@
+	rootMap common.MapStr,
+	path []string,
+	dottedPath string,
+)
```
path and dotted path are the same. How about:

```
type path struct {

urso commented on heartbeat/skima/walk.go:
```diff
@@ -0,0 +1,74 @@
+func walkFull(m common.MapStr, root common.MapStr, path []string, wo walkObserver) {
+	for k, v := range m {
+		splitK := strings.Split(k, ".")
+		var newPath []string
```
minor optimisation:

```
newPath := make([]string, len(path) + len(splitK)) // pre-alloc slice
copy(newPath, path)
copy(newPath[len(path):], splitK)
```

andrewvc replied:
Ah, yes, much nicer

andrewvc commented on heartbeat/skima/core.go:
```diff
@@ -0,0 +1,168 @@
+			currentMap common.MapStr,
+			rootMap common.MapStr,
+			path []string,
+			dottedPath string) {
```
Thanks for the heads up, glad to use a struct. I think that will make it much easier to read

urso commented on heartbeat/skima/walk.go:
```diff
@@ -0,0 +1,74 @@
+		} else if convertedM, ok := v.(Map); ok {
+			mapV = common.MapStr(convertedM)
+			vIsMap = true
+		}
```
also add `map[string]interface{}` ?

e.g.

```
vIsMap := true
switch tmp := v.(type) {
```

andrewvc commented on heartbeat/skima/walk.go:
```diff
@@ -0,0 +1,74 @@
+		// Walk nested maps
+		vIsMap := false
+		var mapV common.MapStr
+		// Note that we intentionally do not handle StrictMap
```
I've killed StrictMap for now, will remove this comment.

urso commented on heartbeat/testutil/hbtestutil.go:
```diff
@@ -0,0 +1,79 @@
+	"github.com/elastic/beats/heartbeat/skima"
+)
+
+var HelloWorldBody = "hello, world!"
```
If you don't want anybody mess with globals, use const here ;)

urso commented on heartbeat/testutil/hbtestutil.go:
```diff
@@ -0,0 +1,79 @@
+	return uint16(p), nil
+}
+
+func MonitorChecks(id string, host string, ip string, scheme string, status string) skima.SchemaValidator {
```
alternative syntax:

```
func MonitorChecks(id, host, ip, scheme, status string) skima.Validator {
...
}
```

urso commented on heartbeat/testutil/hbtestutil.go:
```diff
@@ -0,0 +1,79 @@
+	return skima.Schema(skima.Map{
+		"monitor": skima.Map{
+			// TODO: This is only optional because, for some reason, TCP returns
+			// this value, but HTTP does not. We should fix this
```
I'm not sure if this was actually on purpose. Does http add an `url` field?

andrewvc replied:
HTTP does not, but TCP does, so we should probably be consistent here.

I kind of like having an URL representation of every resource TBH, but then it is one more field to be indexed.... 

WDYT?

urso commented on heartbeat/monitors/active/http/http_test.go:
```diff
@@ -0,0 +1,111 @@
+		},
+		"http": skima.Map{
+			"url": urlStr,
+			// TODO: This should work in the future "response.status_code":   statusCode,
```
Bug in HB. the http monitor returns to early on error, such that the status code is not reported.

ruflin replied:
Got it, would be good to have this note in the code somehow to make it clear.

Can be in a follow up PR.

urso commented on heartbeat/monitors/active/http/task.go:
```diff
@@ -218,41 +218,62 @@ func execPing(
 	timeout time.Duration,
 	validator func(*http.Response) error,
-) (time.Time, time.Time, common.MapStr, reason.Reason) {
+) (start time.Time, end time.Time, event common.MapStr, errReason reason.Reason) {
```
can also be written `(start, end time.Time, event common.MapStr, err reason.Reason)`

urso commented on heartbeat/monitors/active/http/task.go:
```diff
@@ -218,41 +218,62 @@ func execPing(
-	defer resp.Body.Close()
 
 	err = validator(resp)
 	end = time.Now()
```
The different network layers record round trip times. For http having a response doesn't mean the request if finished yet. It means we have seen and parsed the headers. All additional contents will be streamed. If the body is bigger + if the validator consumes all contents, we at least have some idea about the total time taken (the validator generates additional IO). This can definitely be improved, but as we create IO when consuming the body, we need to report durations.

ruflin replied:
thanks for the details.

jsoriano commented on heartbeat/hbtest/skima/results.go:
```diff
@@ -0,0 +1,98 @@
+	kontinue := true
+	for path, pathResults := range r {
+		for _, result := range pathResults {
+			kontinue = f(path, result)
```
```
if !f(path, result) {
    return
}
```
instead of `kontinue`?

andrewvc replied:
Lol, I forgot `return` was a thing. Good suggestion :-D

andrewvc commented:
@jsoriano I think some of the fixes I pushed blew-away your comment and the discussion about merging `skima` and `schema` since the lines in the code changed.

It seems like, as you said, `schema` is about `map`->`event`, and this is more about map validation. While there are some similarities (both walk maps for instance), it seems that the actual goals, as you mentioned are quite different. Merging them doesn't seem like a clear win to me at this point. 

I do completely agree that if we do keep them separate the naming would be confusing. I'm glad to rename `skima` to something else. Maybe `validatron` or `mapval` or `treecomp` etc. etc.

andrewvc commented:
I think I've addressed all PR comments except the question of what to do with `skima` itself.

That aside, I do think this PR is pretty much done. The original intent here was to write some basic tests for http and TCP. I do think more tests are needed, however, I don't think this PR needs more going on inside of it. My preference would be to merge this sooner rather than later, and add those in a separate PR focused on testing the correctness of the various jobs.

Once we have tests for the current behavior, we'll also have a catalog of where consistency can be improved (things like status codes in http responses etc.) and can make some further follow-up PRs to rectify those issues.

Moving in phases for a refactor like this is important IMHO, since I don't want to change/break behavior until we have a way to reflect that in the tests.

jsoriano commented:
I think that `skima` could find its place under libbeat, probably other beats can also make use of it, specially if we move more integration testing towards Go implementations.
Regarding the name, to avoid confusion with `schema`, I'd highlight its use as map validator, so `mapval` or something like this looks great to me, maybe it could be placed in `libbeat/testing/mapval`, or if it can find uses out of testing in `libbeat/common/mapval`. Once this is solved I agree with merging this so development can continue.

ruflin commented:
As these 2 packages are currently in 2 different Beats I don't think the naming discussion should hold back the merging of this PR. @andrewvc Could you open a follow up Github issue for the discussion?

ruflin commented:
@andrewvc Can you look into the failing http tests? Also makes sure to rebase on master, this should resolve the failing tests in the generator.

jsoriano commented:
@ruflin `schema` is in `libbeat` and it is currently used both in libbeat and Metricbeat, and I guess it could be also used in heartbeat (or other beats too) if needed at some point. The same for "`skima`", it has value to be used in other beats tests, so apart of the naming, wdyt about moving it to libbeat?

(In any case I'd prefer a name that highlights its use for validation :slightly_smiling_face: like `mapval` or something like this)

ruflin commented:
@jsoriano Sorry, you are right. I forgot that we move schema to libbeat :-( I'm all on board with changing the name but I wonder if it must happen in this PR. The PR already got much larger then initially expected so I would rather get it in and have a follow up PR to changes the names and potentially moves the package around.

andrewvc commented:
OK, renamed `skima->mapval`. Also, moved `mapval` to libbeat.

I also fixed the source of the test failure, which was due to a restructuring I made to the HTTP task. This actually fixes what I thought was a pre-existing bug, but was actually something I introduced, which was HTTP error responses not including their status code in the event. The good news is that adding these tests will help clarify such things in the future.

andrewvc commented:
@ruflin this is good for a squash + merge FYI. I'm leaving it unsquashed for now in case you want to look over the last few commits.

ruflin commented:
Thanks for all the work on this. Solid foundation for more testing on Heartbeat.
